zerver/lib/stream_subscription.py

```diff
@@ -313,3 +313,19 @@ def get_subscriptions_for_send_message(
         )
     )
     return query
+
+
+def update_stream_subscriber_count(
+    stream_to_subscribers: dict[Stream, set[UserProfile]], count_sign: Literal[1, -1]
+) -> None:
+    """increment/decrement subscriber_count for non-deactivated stream subscribers.
+
+    count_sign -> 1=increment, -1=decrement
+    """
+
+    for stream, subscribers in stream_to_subscribers.items():
+        stream.subscriber_count += (
+            len([user.id for user in subscribers if user.is_active]) * count_sign
+        )
+
+    Stream.objects.bulk_update(stream_to_subscribers.keys(), ["subscriber_count"])
```

Comment:
For further optimization, we can add a new parameter `active_subscribers` which callers can pass as `True` if they are sure that all passed subscribers are already active, then we won't need to check for `user.is_active` in the loop

Reply:
I would also look hard at whether we can do a database query that asks the database to do the deltas; that might avoid certain classes of races?